// whatap/lib/pack/packenum.js
/**
 * Copyright 2016 the WHATAP project authors. All rights reserved.
 * Use of this source code is governed by a license that
 * can be found in the LICENSE file.
 */

exports.PARAMETER   = 0x0100;
exports.COUNTER_1   = 0x0201;
exports.PROFILE     = 0x0300;
exports.ACTIVESTACK_1 = 0x0401;
exports.TEXT        = 0x0700;
exports.ERROR_SNAP  = 0x0800;
exports.ERROR_SNAP_1  = 0x0801;
exports.REALTIME_USER_1 = 0x0f01;

exports.STAT_SERVICE= 0x0900;
exports.STAT_SERVICE_1 = 0x0901;
exports.STAT_GENERAL= 0x0910;
exports.STAT_SQL    = 0x0a00;
exports.STAT_HTTPC  = 0x0b00;
exports.STAT_ERROR  = 0x0c00;
exports.STAT_METHOD = 0x0e00;
exports.STAT_TOP_SERVICE= 0x1000;
exports.STAT_REMOTE_IP  = 0x1100;
exports.STAT_USER_AGENT = 0x1200;

exports.EVENT       = 0x1400;
exports.HITMAP      = 0x1500;
exports.HITMAP_1    = 0x1501;
exports.COUNTER5M   = 0x1600;
exports.HITMAP5M    = 0x1700;

exports.TAG_COUNT   = 0x1601;
exports.ZIP         = 0x170b;

exports.PACK_NAME = {
    0x0100 : 'PARAMETER',
    0x0201 : 'COUNTER_1',
    0x0300 : 'PROFILE',
    0x0401 : 'ACTIVESTACK_1',
    0x0700 : 'TEXT',
    0x0800 : 'ERROR_SNAP',
    0x0801 : 'ERROR_SNAP_1',
    0x0f01 : 'REALTIME_USER_1',
    0x0900 : 'STAT_SERVICE',
    0x0901 : 'STAT_SERVICE_1',
    0x0910 : 'STAT_GENERAL',
    0x0a00 : 'STAT_SQL',
    0x0b00 : 'STAT_HTTPC',
    0x0c00 : 'STAT_ERROR',
    0x0e00 : 'STAT_METHOD',
    0x1000 : 'STAT_TOP_SERVICE',
    0x1100 : 'STAT_REMOTE_IP',
    0x1200 : 'STAT_USER_AGENT',
    0x1400 : 'EVENT',
    0x1500 : 'HITMAP',
    0x1501 : 'HITMAP_1',
    0x1600 : 'COUNTER5M',
    0x1700 : 'HITMAP5M',
    0x1601 : 'TAG_COUNT',
    0x170b : 'ZIP'
};
